jenkins/dataproc-custom-images/prepare-custom-leonardo-jupyter-dataproc-image.sh

```diff
@@ -0,0 +1,186 @@
+# the entry must match the var name above, which must correspond to a valid docker URI
+docker_image_var_names="welder_server leonardo_jupyter terra_jupyter_base terra_jupyter_python openidc_proxy"
+
+#
+# Functions
+#
+
+# Retry a command up to a specific number of times until it exits successfully,
+# with exponential back off.
+#
+# $ retry 5 echo "Hello"
+# Hello
+#
+# $ retry 5 false
+# Retry 1/5 exited 1, retrying in 2 seconds...
+# Retry 2/5 exited 1, retrying in 4 seconds...
+# Retry 3/5 exited 1, retrying in 8 seconds...
+# Retry 4/5 exited 1, retrying in 16 seconds...
+# Retry 5/5 exited 1, no more retries left.
+#
+function retry {
+    local retries=$1
+    shift
+
+    for ((i = 1; i <= retries; i++))
+    do
+        # run with an 'or' so set -e doesn't abort the bash script on errors
+        exit=0
+        "$@" || exit=$?
+        if [[ $exit -eq 0 ]]; then
+            return 0
+        fi
+        wait=$((2 ** i))
+        if [[ $i -eq $retries ]]; then
+            log "Retry $i/$retries exited $exit, no more retries left."
+            break
+        fi
+        log "Retry $i/$retries exited $exit, retrying in $wait seconds..."
+        sleep $wait
+    done
+    return 1
+}
+
+function log() {
+    printf '[%s]: %s\n' \
+        "$(date +'%Y-%m-%dT%H:%M:%S%z')" \
+        "$*"
+}
+
+#
+# Main
+#
+log 'Installing prerequisites...'
+
+# Obtain the latest valid apt-key.gpg key file from https://packages.cloud.google.com to work
+# around intermittent apt authentication errors. See:
+# https://cloud.google.com/compute/docs/troubleshooting/known-issues
+retry 5 curl https://packages.cloud.google.com/apt/doc/apt-key.gpg | apt-key add -
+retry 5 apt-key update
+
+# install Docker
+# https://docs.docker.com/install/linux/docker-ce/debian/
+# export DOCKER_CE_VERSION="19.03.2~ce~3-0~debian"
+
+# retry 5 betterAptGet
+retry 5 apt-get install -y -q \
+    apt-transport-https \
+    ca-certificates \
+    curl \
+    gnupg2 \
+    software-properties-common
+
+log 'Adding Docker package sources...'
+
+apt-get remove docker docker-engine
+
+# shellcheck disable=SC1091
+os_dist_name="$(. /etc/os-release; echo "$ID")"
+os_dist_code_name="$(lsb_release -cs)"
+os_dist_release_channel="stable"
+os_dist_arch="amd64"
+
+docker_gpg_key_url="https://download.docker.com/linux/${os_dist_name:?}/gpg"
+docker_apt_repo_url="https://download.docker.com/linux/${os_dist_name:?}"
+
+retry 5 curl -fsSL "${docker_gpg_key_url:?}" | apt-key add -
+
+add-apt-repository \
+  "deb [arch=${os_dist_arch:?}] ${docker_apt_repo_url:?} \
+  ${os_dist_code_name:?} \
+  ${os_dist_release_channel:?}"
+
+log 'Installing Docker...'
+
+retry 5 apt-get update
+retry 5 add-apt-repository "deb [arch=amd64] https://download.docker.com/linux/debian $(lsb_release -cs) stable"
+
+dpkg --configure -a
+#this line fails consistently, but it does not fail in a fatal way so we add `|| true` to prevent the script from halting execution
+apt-get install -y -q docker-ce || true
```

Comment:
This is the line I got hung up on for several days.... I've tried everything I can think of from changing the debian version, to the docker-ce version, to messing with systemctl docker and the docker installation folders. All attempts failed with the error detailed here https://fc-jenkins.dsp-techops.broadinstitute.org/job/leonardo-dataproc-image-2/65/console.

I ended up putting a long sleep before this line, SSH'd into the VM, and found that even though this line fails, `systemctl status service.docker` returns a healthy status. I decided to just `|| true` and ignore the failure and everything seems to be working......

Reply:
This is quite strange, I haven't seen this before. I'd like to try it out myself to reproduce...

Reply:
Did `-f` do anything? If not I'm fine with leaving `|| true` here, as discussed in person.

Reply:
it did not unfortunately